src/main/java/org/opensearch/knn/index/MethodComponentContext.java

```diff
@@ -36,13 +38,18 @@
  *
  * Each component is composed of a name and a map of parameters.
  */
-@AllArgsConstructor
+@RequiredArgsConstructor
 public class MethodComponentContext implements ToXContentFragment, Writeable {
 
     @Getter
     private final String name;
     private final Map<String, Object> parameters;
 
+    // Need a code over here to make sure that we are reading and writing in stream this optional index version.
```

Comment:
This seems potentially error prone. Im wondering if it makes sense to build into the streams/parsing logic.

Reply:
I thought about it a lot. and went back and forth on that. But then I realized we don’t need to put it in the stream because indexCreatedVersion is not coming from user so we will not be parsing that information. indexCreatedVersion is an attribute we are setting while creating the mapper. So, its is not required to send or read from stream

Reply:
A couple things for this. 

First (unrelated to your change, but just want to add here so I dont forget), it looks like the streaminput logic here is wrong: https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/MethodComponentContext.java#L55-L62. Just because `in.available()` is not empty, does not mean those bytes are for the parameters. Ill fix this later.

Second, the stream operators are used during training to pass the modelcontext to other nodes: https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/plugin/rest/RestTrainModelHandler.java#L98. During training, I am guessing defaults are going to be looked up in some fashion when building the model (think HNSWPQ): https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/util/Faiss.java#L183. So we may need to set this either on the node that is selected for training or add it to the stream.

Third, in the future, I think we will want to store the model context in the cluster. We may want to eventually store this information there as well - for instance, if we decide to allow some model parameters to be dynamically adjustable (ef_const for HNSWPQ for instance).

Reply:
@jmazanec15 
Thanks for pointing out. This was a little tricky.

> First (unrelated to your change, but just want to add here so I dont forget), it looks like the streaminput logic here is wrong: https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/MethodComponentContext.java#L55-L62. Just because in.available() is not empty, does not mean those bytes are for the parameters. Ill fix this later.

I also looked at this and realized that it was not a good approach but then I left it as it is because the code is working. :)

> Second, the stream operators are used during training to pass the modelcontext to other nodes: https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/plugin/rest/RestTrainModelHandler.java#L98. During training, I am guessing defaults are going to be looked up in some fashion when building the model (think HNSWPQ): https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/util/Faiss.java#L183. So we may need to set this either on the node that is selected for training or add it to the stream.

Actually I would still say we don't need to put index version in streams. We can go around that only by setting the index created version when the Training job is getting started which is here: https://github.com/opensearch-project/k-NN/blob/33da521e0f98317b4700b62807e1d21b11f54a71/src/main/java/org/opensearch/knn/training/TrainingJob.java#L180

There are other places also where we can set this information:
1. In Validate function of TrainingModelRequest: https://github.com/opensearch-project/k-NN/blob/98b62183c878cc0334453ebc6947cb9424991658/src/main/java/org/opensearch/knn/plugin/transport/TrainingModelRequest.java#L301-L307
2. In the RestTrainModelHandler somewhere in around this: https://github.com/opensearch-project/k-NN/blob/98b62183c878cc0334453ebc6947cb9424991658/src/main/java/org/opensearch/knn/plugin/rest/RestTrainModelHandler.java#L98

In the #1 the problem is its the validate function, and we should not update objects in validate functions. 

#2 is the place where we have just parsed the request so we add the indexCreatedVersion there we need to validate few more things like if training index is created or not. These validation are already happening in the function mentioned #1.

So based on these reason I think we should just put the index created version just before starting the job which is here: https://github.com/opensearch-project/k-NN/blob/33da521e0f98317b4700b62807e1d21b11f54a71/src/main/java/org/opensearch/knn/training/TrainingJob.java#L180


> Third, in the future, I think we will want to store the model context in the cluster. We may want to eventually store this information there as well - for instance, if we decide to allow some model parameters to be dynamically adjustable (ef_const for HNSWPQ for instance).

This is a future use case and not sure how much valid it is. We should look at this one once it arrives.


The reason I am more pushing back on not setting these values in streams is because setting info in streams comes with version checks which make the code less maintainable. Hence I am trying to avoid these checks and see if we can get around those in some fashion. If needed we can always add the values in the streams when needed. Right now I am not seeing a usecase.

Reply:
I think setting before the training job makes sense. 

The only thing that is a bit concerning is if someone uses the ModelContext in the future and forgets to set the versioning information. Then, tries to use the version in the code. Not necessarily for this review, but do you have any ideas for mitigating this risk?